### DISCUSSION TITLE
Correct way to use `mpc.np_concatenate()`

### DISCUSSION THREAD

**Discussion by abraj** [lschoe/mpyc#100] [Q&A] [answered]
I am getting the error `IndexError: tuple index out of range` for the following simple code snippet:
```python
secint = mpc.SecInt(32)

arr1 = np.array([secint(1), secint(2), secint(3)])
arr2 = np.array([secint(4), secint(5), secint(6)])

concat_arr = mpc.np_concatenate((arr1, arr2))  # IndexError: tuple index out of range
print(concat_arr)
```

Am I not using [`mpc.np_concatenate()`](https://mpyc.readthedocs.io/en/latest/mpyc.html#mpyc.runtime.Runtime.np_concatenate) correctly?

**Comment by lschoe** [comment, marked as answer]
Indeed, you need to do things a bit differently.

To start with, do not use the "internal/low-level" function `mpc.np_concatenate()` in your code at user level. Instead, use the standard Numpy call, like this:
```python
concat_arr = np.concatenate((arr1, arr2))
```
The MPyC runtime will figure out (under the hood) what to do if any secure arrays are involved. However, arrays `arr1`and `arr2` as you've defined them aren't secure arrays: these are just arrays that happen to have `secint` elements.

So, what you probably want to write is this:
```python
arr1 = secint.array(np.array([1, 2, 3]))
arr2 = secint.array(np.array([4, 5, 6]))

concat_arr = np.concatenate((arr1, arr2))
```
If you print this `concat_arr` you get something like `<mpyc.sectypes.ArraySecInt32 object at 0x000001A67053ABC0>`.
To see the value in the clear do something like `print(await mpc.ouput(concat_arr))`.

So, the contents of array `concat_arr` is secret-shared, but its shape is public, and you can look it up as for any Numpy array using `concat_arr.shape` which gives `(6,)` as result.

Take a look at demos like [np_id3gini.py](../blob/master/demos/np_id3gini.py) and  [np_lpsolver.py](../blob/master/demos/np_lpsolver.py) and so on to see the use of secure Numpy arrays "in action".

**Reply by abraj** [reply]
Got it! Thanks.

Btw, are you suggesting that all `mpc.np_*` functions are "internal/low-level", and we should avoid using them, if possible.
What about array attributes like `z.T`? (Is it okay to use them?)
I mean, should `np.transpose(z)` be used as opposed to `z.T` or `mpc.np_transpose(z)` ?

**Reply by lschoe** [reply]
Sure, that works. You can try it out yourself, but you can also see it used for example in the demo for computing the Moore-Penrose pseudoinverse: https://github.com/lschoe/mpyc/blob/15f747c0c7ff849b9a8a2a3f4e9d9439ab547b6e/demos/pseudoinverse.py#L63-L78
Using `np.transpose(z)` is also fine, but `mpc.np_transpose(z)` is low-level/internal MPyC code.

**Reply by abraj** [reply]
Thanks.
